pdg/src/builder.rs

```diff
@@ -231,6 +231,13 @@ pub fn add_node(
         .unwrap_or_else(|| graphs.graphs.push(Graph::new()));
     let node_id = graphs.graphs[graph_id].nodes.push(node);
 
+    // Mark the graph for null pointers
+    let ptr_is_null = ptr.map_or(false, |ptr| ptr == 0);
+    let graph_is_null = &mut graphs.graphs[graph_id].is_null;
+    // Assert that we're not mixing null and non-null pointers
+    assert!(graph_is_null.is_none() || *graph_is_null == Some(ptr_is_null));
+    *graph_is_null = Some(ptr_is_null);
```

Comment:
I wonder if the `Option` for `is_null` can be removed entirely.  It seems like it only exists so `Graph::new()` has a default value that's non-conflicting with both `true` and `false`.  But if this is the only place where new graphs are created, then we could pass `ptr_is_null` to `Graph::new()` and initialize the `is_null` field directly to the correct value.